File: itx-go/itx-backend/internal/state/props.go
package state

import (
	_ "embed"
	"fmt"

	"github.com/chehsunliu/itx/itx-go/itx-impl/config"
	"github.com/chehsunliu/itx/itx-go/itx-impl/queue/rabbitmq"
	"github.com/chehsunliu/itx/itx-go/itx-impl/queue/sqs"
	"github.com/chehsunliu/itx/itx-go/itx-impl/repo/mariadb"
	"github.com/chehsunliu/itx/itx-go/itx-impl/repo/postgres"
	"gopkg.in/yaml.v3"
)

//go:embed application.yaml
var rawApplicationYAML string

type AppProps struct {
	DBProvider    string                    `yaml:"db-provider"`
	QueueProvider string                    `yaml:"queue-provider"`
	MariaDB       mariadb.RepoFactoryProps  `yaml:"mariadb"`
	Postgres      postgres.RepoFactoryProps `yaml:"postgres"`
	RabbitMQ      rabbitmq.FactoryProps     `yaml:"rabbitmq"`
	SQS           sqs.FactoryProps          `yaml:"sqs"`
}

func PropsFromEnv() (AppProps, error) {
	var props AppProps
	if err := yaml.Unmarshal([]byte(config.SubstituteEnv(rawApplicationYAML)), &props); err != nil {
		return AppProps{}, fmt.Errorf("parse application.yaml: %w", err)
	}
	return props, nil
}

File: itx-rs/itx-backend/src/state/props.rs
use itx_impl::queue::rabbitmq::factory::RabbitMessageQueueFactoryProps;
use itx_impl::queue::sqs::factory::SqsMessageQueueFactoryProps;
use itx_impl::repo::mariadb::factory::MariaDbRepoFactoryProps;
use itx_impl::repo::postgres::factory::PostgresRepoFactoryProps;
use serde::Deserialize;
use std::error::Error;

#[derive(Clone, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct AppProps {
    pub db_provider: Option<String>,
    pub queue_provider: Option<String>,
    pub mariadb: MariaDbRepoFactoryProps,
    pub postgres: PostgresRepoFactoryProps,
    pub rabbitmq: RabbitMessageQueueFactoryProps,
    pub sqs: SqsMessageQueueFactoryProps,
}

impl AppProps {
    pub fn from_env() -> Result<AppProps, Box<dyn Error>> {
        let raw_props_content = include_str!("./application.yaml");
        let props_content = subst::substitute(raw_props_content, &subst::Env)?;
        let props = serde_yaml::from_str::<AppProps>(&props_content)?;
        Ok(props)
    }
}
